ion-schema/src/nfa.rs

```diff
@@ -0,0 +1,258 @@
+use crate::ion_path::IonPath;
+use crate::system::{TypeId, TypeStore};
+use crate::types::TypeValidator;
+use crate::IonSchemaElement;
+use ion_rs::value::owned::Element;
+use std::collections::{HashMap, HashSet};
+use std::rc::Rc;
+
+/// Represents an event in NFA which can either be an Ion Value or an end of stream
+#[derive(Debug, Clone, PartialEq)]
+pub enum Event {
+    Value(Element),
+    EndOfStream,
+}
+
+impl Event {
+    // Note: can not use this as a From<Element> implementation as Element is not defined within this crate
+    pub(crate) fn from_elements_to_events<T: IntoIterator<Item = Element>>(iter: T) -> Vec<Event> {
+        let mut events: Vec<Event> = iter
+            .into_iter()
+            .map(<Element as Into<Event>>::into)
+            .collect();
+        // add END_OF_STREAM at the end of events vec
+        events.push(Event::EndOfStream);
+        events
+    }
+}
+
+impl From<Element> for Event {
+    fn from(value: Element) -> Self {
+        Event::Value(value)
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct Transition {
+    destination: usize,      // represents destination state
+    to_optional_state: bool, // represents if destination state is an optional state or not
```

Comment:
I don't understand why `to_optional_state` is necessary. Shouldn't all of that be handled by the min occurs for the state itself?

Reply:
Yes, I can remove this.